File: windows/WeChatExporter.Windows/Services/DatImageDecoder.cs
using System.Diagnostics;
using System.IO;

namespace WeChatExporter.Services;

/// <summary>解密微信 .dat 图片，优先调用 wx-cli，失败时尝试 XOR 探测。</summary>
internal static class DatImageDecoder
{
    public static async Task<int> DecodeDatFilesAsync(string outputDir, Action<string> log, CancellationToken cancellationToken = default)
    {
        var mediaRoot = Path.Combine(outputDir, "media");
        if (!Directory.Exists(mediaRoot)) return 0;

        var datFiles = Directory.EnumerateFiles(mediaRoot, "*.dat", SearchOption.AllDirectories).ToList();
        if (datFiles.Count == 0) return 0;

        var decoded = 0;
        var accountDir = LocateWeChatAccountDir();
        var wxCli = WxCliService.LocateExecutable();

        foreach (var dat in datFiles)
        {
            var baseName = Path.GetFileNameWithoutExtension(dat);
            var outDir = Path.GetDirectoryName(dat)!;
            if (ExistingDecodedImage(baseName, outDir) is not null)
            {
                decoded++;
                continue;
            }

            if (accountDir is not null && wxCli is not null && await DecodeWithWxCliAsync(dat, accountDir, wxCli, log, cancellationToken))
            {
                decoded++;
                continue;
            }

            if (DecodeWithXor(dat))
            {
                decoded++;
                log($"已解密图片（XOR）：{Path.GetFileName(dat)}");
            }
        }

        if (decoded > 0) log($"已解密 {decoded} 张 .dat 图片");
        return decoded;
    }

    public static byte[]? TryDecodeInline(byte[] data)
    {
        for (var key = 0; key <= 255; key++)
        {
            var sample = XorDecode(data, (byte)key);
            if (ImageExporter.SniffImageMime(sample) is not null) return sample;
        }
        return null;
    }

    private static string? ExistingDecodedImage(string baseName, string dir)
    {
        foreach (var ext in new[] { "jpg", "jpeg", "png", "gif", "webp" })
        {
            var path = Path.Combine(dir, $"{baseName}.{ext}");
            if (File.Exists(path)) return path;
        }
        return null;
    }

    private static async Task<bool> DecodeWithWxCliAsync(
        string datPath, string accountDir, string wxCli, Action<string> log, CancellationToken cancellationToken)
    {
        var outPath = Path.Combine(Path.GetDirectoryName(datPath)!, $"{Path.GetFileNameWithoutExtension(datPath)}.jpg");
        var psi = new ProcessStartInfo
        {
            FileName = wxCli,
            Arguments = $"decode-image \"{datPath}\" -d \"{accountDir}\" -o \"{outPath}\"",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        try
        {
            using var process = Process.Start(psi);
            if (process is null) return false;
            await process.WaitForExitAsync(cancellationToken);
            if (File.Exists(outPath) && new FileInfo(outPath).Length > 0)
            {
                log($"已解密图片：{Path.GetFileName(datPath)}");
                return true;
            }
        }
        catch
        {
            return false;
        }
        return false;
    }

    private static bool DecodeWithXor(string datPath)
    {
        var data = File.ReadAllBytes(datPath);
        var decoded = TryDecodeInline(data);
        if (decoded is null || ImageExporter.SniffImageMime(decoded) is not { } mime) return false;

        var ext = mime switch
        {
            "image/png" => "png",
            "image/gif" => "gif",
            "image/webp" => "webp",
            _ => "jpg"
        };
        var outPath = Path.Combine(Path.GetDirectoryName(datPath)!, $"{Path.GetFileNameWithoutExtension(datPath)}.{ext}");
        File.WriteAllBytes(outPath, decoded);
        return true;
    }

    private static byte[] XorDecode(byte[] data, byte key) => data.Select(b => (byte)(b ^ key)).ToArray();

    public static string? LocateWeChatAccountDir()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var roots = new[]
        {
            Path.Combine(home, "Documents", "xwechat_files"),
            Path.Combine(home, "xwechat_files"),
        };

        string? best = null;
        var bestTime = DateTime.MinValue;
        foreach (var root in roots.Where(Directory.Exists))
        {
            foreach (var entry in Directory.EnumerateDirectories(root))
            {
                if (Path.GetFileName(entry) == "all_users") continue;
                var msgDir = Path.Combine(entry, "msg");
                if (!Directory.Exists(msgDir)) continue;
                var time = Directory.GetLastWriteTimeUtc(entry);
                if (time <= bestTime) continue;
                bestTime = time;
                best = entry;
            }
        }
        return best;
    }
}

File: Sources/WeChatExporter/Services/DatImageDecoder.swift
import CommonCrypto
import Foundation

/// 解密微信 .dat 图片，优先调用 wx-cli，失败时尝试简单 XOR 探测。
enum DatImageDecoder {
    static func decodeDatFiles(in outputDir: URL, log: @escaping (String) -> Void) async -> Int {
        let mediaRoot = outputDir.appendingPathComponent("media", isDirectory: true)
        guard FileManager.default.fileExists(atPath: mediaRoot.path) else { return 0 }

        var datFiles: [URL] = []
        if let enumerator = FileManager.default.enumerator(
            at: mediaRoot,
            includingPropertiesForKeys: [.isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) {
            for case let fileURL as URL in enumerator {
                guard (try? fileURL.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) == true else { continue }
                if fileURL.pathExtension.lowercased() == "dat" {
                    datFiles.append(fileURL)
                }
            }
        }

        guard !datFiles.isEmpty else { return 0 }

        var decoded = 0
        let accountDir = locateWeChatAccountDir()
        let wxCli = WxCliService.locateExecutable()

        for dat in datFiles {
            let base = dat.deletingPathExtension().lastPathComponent
            let outDir = dat.deletingLastPathComponent()
            if existingDecodedImage(base: base, in: outDir) != nil {
                decoded += 1
                continue
            }

            if let accountDir, let wxCli,
               await decodeWithWxCli(dat: dat, accountDir: accountDir, wxCli: wxCli, log: log) {
                decoded += 1
                continue
            }

            if decodeWithXor(dat: dat) {
                decoded += 1
                log("已解密图片（XOR）：\(dat.lastPathComponent)")
            }
        }

        if decoded > 0 {
            log("已解密 \(decoded) 张 .dat 图片")
        }
        return decoded
    }

    static func tryDecodeInline(_ data: Data) -> Data? {
        for key in 0...255 {
            let sample = xorDecode(data, key: UInt8(key))
            if ImageExporter.sniffImageMIME(sample) != nil {
                return sample
            }
        }
        return nil
    }

    private static func existingDecodedImage(base: String, in dir: URL) -> URL? {
        for ext in ["jpg", "jpeg", "png", "gif", "webp"] {
            let url = dir.appendingPathComponent("\(base).\(ext)")
            if FileManager.default.fileExists(atPath: url.path) { return url }
        }
        return nil
    }

    private static func decodeWithWxCli(
        dat: URL,
        accountDir: URL,
        wxCli: URL,
        log: @escaping (String) -> Void
    ) async -> Bool {
        let outBase = dat.deletingPathExtension().lastPathComponent
        let outURL = dat.deletingLastPathComponent().appendingPathComponent("\(outBase).jpg")
        let process = Process()
        process.executableURL = wxCli
        process.arguments = ["decode-image", dat.path, "-d", accountDir.path, "-o", outURL.path]
        let errPipe = Pipe()
        process.standardOutput = Pipe()
        process.standardError = errPipe
        do {
            try process.run()
            process.waitUntilExit()
            if FileManager.default.fileExists(atPath: outURL.path),
               (try? Data(contentsOf: outURL))?.isEmpty == false {
                log("已解密图片：\(dat.lastPathComponent)")
                return true
            }
        } catch {
            return false
        }
        return false
    }

    @discardableResult
    private static func decodeWithXor(dat: URL) -> Bool {
        guard let data = try? Data(contentsOf: dat), !data.isEmpty else { return false }
        guard let decoded = tryDecodeInline(data),
              let mime = ImageExporter.sniffImageMIME(decoded) else { return false }
        let ext: String
        switch mime {
        case "image/png": ext = "png"
        case "image/gif": ext = "gif"
        case "image/webp": ext = "webp"
        default: ext = "jpg"
        }
        let outURL = dat.deletingLastPathComponent().appendingPathComponent("\(dat.deletingPathExtension().lastPathComponent).\(ext)")
        do {
            try decoded.write(to: outURL, options: .atomic)
            return true
        } catch {
            return false
        }
    }

    private static func xorDecode(_ data: Data, key: UInt8) -> Data {
        Data(data.map { $0 ^ key })
    }

    static func locateWeChatAccountDir() -> URL? {
        let home = FileManager.default.homeDirectoryForCurrentUser
        let roots = [
            home.appendingPathComponent("Library/Containers/com.tencent.xinWeChat/Data/Documents/xwechat_files", isDirectory: true),
            home.appendingPathComponent("Documents/xwechat_files", isDirectory: true),
            home.appendingPathComponent("xwechat_files", isDirectory: true),
        ]

        var candidates: [(URL, Date)] = []
        for root in roots where FileManager.default.fileExists(atPath: root.path) {
            guard let entries = try? FileManager.default.contentsOfDirectory(
                at: root,
                includingPropertiesForKeys: [.contentModificationDateKey],
                options: [.skipsHiddenFiles]
            ) else { continue }
            for entry in entries where entry.hasDirectoryPath && entry.lastPathComponent != "all_users" {
                let msgDir = entry.appendingPathComponent("msg", isDirectory: true)
                if FileManager.default.fileExists(atPath: msgDir.path) {
                    let date = (try? entry.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
                    candidates.append((entry, date))
                }
            }
        }

        return candidates.max(by: { $0.1 < $1.1 })?.0
    }
}
